nemo_aligner/utils/utils.py

```diff
@@ -365,3 +368,17 @@ def convert_to_amp_o2_format(state_dict):
         new_state_dict[new_key] = state_dict[key]
 
     return new_state_dict
+
+
+def make_sharded_optimizer_tensor_and_state(model_param, optim_param, prefix: str):
```

Comment:
This function looks pretty similar to https://github.com/NVIDIA/Megatron-LM/blob/548e57a00d985c15d5c56fb4749656cf011be9ee/megatron/core/dist_checkpointing/optimizer.py#L66 except that it also handles `ShardedObject`.

Could you please:
1. Add a comment to mention this and explain why we need to handle `ShardedObject` here, so that someone who reads this code has a clue of what this is about.
2. Rename this function and its parameters (and update the error message in the assert above) so that they don't refer to some optimizer? (since we're not using it with an optimizer). They could instead refer to a reference policy but this is a generic function so it could also just be something like `other_param`.

Reply:
Done

Reply:
Thanks -- sorry to be annoying, but I'm afraid the way you named parameters is confusing because `reference_param` is actually not used for the parameters of the reference policy. As far as I can understand the code is correct and it's just that "reference" may actually refer to different concepts, which is likely to confuse people looking at this code. Could you please rename again? (for instance you could use "base" or "template" instead of "reference")

Reply:
Sorry if it's confusing, but the parameter names now are correct to the logic of what is actually occuring in the function. We read the TP/PP/DP axis information from the `reference_param` and use that information so save down (via dataclasses.replace) the `model_param`. This is what the function does and how it used to save the ref policy in SPIN. The reference_param input is not being changed, it is being used as a reference for the TP/PP/DP axis information in order to correctly save in sharded format. If you don't have a reference object to read this from, then your model won't save the correct TP/PP/DP axis.

Ergo, the function now has correct param names in the spirit of what it is actually doing. It has nothing to do with the reference policy in SPIN, that's just one algo which uses this function.

Reply:
Yeah, I understand that, I'm just saying this is quite confusing to someone reading the code and I was thus suggesting to rename the parameters to avoid this confusion.

If you insist on keeping the current names can you please at least rephrase this comment
```
# (...) can be copied from some
# reference model, like in the case of a reference policy which has a 1:1 architecture as some actor policy
```
which reinforces this confusion since the "reference model" is not actually the "reference policy" in practice.